### PR TITLE
Support signing using a Google KMS provider

### PR DESCRIPTION
**Pull request by netomi**
This is a draft PR as the deployment is not finished yet.
Need to figure out how to provide the google cloud credentials file and cert chain to the docker image.

cc @mbarbero

### CODE REVIEW

**Comment by mbarbero** [comment]
Great progress already. Awesome work.

As we eventually package the dependencies into a single JAR, would it be worthwhile to add JSign as a dependency in the pom.xml file? This approach would make updates easier to manage compared to the current method of downloading the JAR in the Dockerfile.

**Comment by netomi** [comment]
yeah that would make sense, we just need to reference then the distributed runtime jar in the config file itself.

**Comment by ebourg** [comment]
You probably want to depend on jsign-core instead of the all-in-one jsign artifact. That will allow you to exclude some unnecessary dependencies such as Apache POI. Also once Jsign 7.0 is released you'll be able to replace jsign-core with jsign-crypto, that's a new artifact containing only the JCA provider and not the Authenticode signing stuff.

**Comment by netomi** [comment]
@ebourg ty for the heads up. Currently we use osslsigncode for signing windows executables, but will most likely replace that by using jsign. We just started with the jar signer as this is more pressing for us.

btw. I looked into the reason why you have to add --add-module java.sql and figured its related to the json-io library. Do you have plans to replace that library with something else? While the library works well, its update strategy is a bit confusing, doing breaking API changes in bugfix releases ...

**Comment by ebourg** [comment]
I picked json-io because it's very compact and I try to keep the size of the all-in-one jar as small as possible. But the size of the latest releases has increased and it's now close to Gson. Since the needs of Jsign are very basic I'm considering replacing it with a custom JSON parser/formatter to save a few kilobytes.

File: webservice/signing/jar/src/main/java/org/eclipse/cbi/webservice/signing/jar/JarSigner.java [hunk 1]
```diff
@@ -255,14 +116,23 @@ public Path signJar(Path jar, SignatureAlgorithm sigAlg, MessageDigestAlgorithm
...
 	 */
 	private ImmutableList<String> createCommand(Path jar, SignatureAlgorithm sigAlg, MessageDigestAlgorithm digestAlg, String sigFile) {
-		ImmutableList.Builder<String> command = ImmutableList.<String>builder().add(jarSigner().toString());
-		
-		if (!Strings.isNullOrEmpty(httpProxyHost())) {
-			command.add("-J-Dhttp.proxyHost=" + httpProxyHost()).add("-J-Dhttp.proxyPort=" + httpProxyPort());
+		ImmutableList.Builder<String> command = ImmutableList.<String>builder().add(configuration().getJarSigner().toString());
+
+		if (!Strings.isNullOrEmpty(configuration().getJavaArgs())) {
+			String[] args = configuration().getJavaArgs().split(" ");
```
Comment (mbarbero) [inline webservice/signing/jar/src/main/java/org/eclipse/cbi/webservice/signing/jar/JarSigner.java:122]: I'd use [Guava's splitter](https://github.com/google/guava/wiki/StringsExplained#splitter), a more robust version of String#split, with less deceitful / more predictable results in many edge cases.

**Comment by netomi** [comment]
I needed to split the configuration for jce and default mode as the config is quite different, maybe its possible to make that generic somehow but wanted to have something working asap.

**Comment by netomi** [comment]
btw. I switched to ghcr.io as repository for the docker image for the jar-signing-service such that I can also push an image and pull it from there when deploying it to the kubernetes cluster as I dont have the credentials for docker.io

**Comment by ebourg** [comment]
> I needed to add an exclude filter for *.DSA and *.SF files to get it to work. Did not yet check which dependency is signed to trigger that.

It probably came from Bouncy Castle, the bcprov artifact has to be signed by a certificate issued by Oracle to expose a JCA provider able to perform encryption operations. Jsign doesn't use BC encryption, so it's fine if the signature is stripped from the all-in-one jar.

**Comment by netomi** [comment]
hm learned something new, when rebasing the PR using the UI button, my commits are not signed anymore and in vigilante mode show as unverified. So I squashed again and force pushed.

**Review by mbarbero — APPROVED** [review]

**Pull request merged**